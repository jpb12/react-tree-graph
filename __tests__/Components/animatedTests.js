import React from 'react';
import { mount, shallow } from 'enzyme';

import { easeQuadOut } from 'd3-ease';
import Animated from '../../src/components/animated';

jest.useFakeTimers();

const nodes = [
	{
		x: 1,
		y: 2,
		data: {
			name: 'Colour'
		}
	}, {
		x: 100,
		y: 50,
		data: {
			name: 'Black'
		}
	}
];

const links = [{
	source: nodes[0],
	target: nodes[1]
}];

const defaultProps = {
	animated: true,
	getChildren: n => n.children,
	height: 100,
	width: 100,
	keyProp: 'name',
	labelProp: 'name',
	nodeOffset: 1,
	nodeRadius: 1,
	duration: 1,
	easing: easeQuadOut,
	links: links,
	nodes: nodes,
	steps: 1,
	circleProps: {},
	gProps: {},
	pathProps: {},
	svgProps: {},
	textProps: {}
};

describe('<Animated>', () => {
	test('renders correctly and sets initial state', () => {
		const tree = shallow(<Animated {...defaultProps}/>);
		expect(tree).toMatchSnapshot();

		expect(tree.state().nodes[1].x).toBe(1);
		expect(tree.state().nodes[1].y).toBe(2);

		expect(tree.state().links[0].target.x).toBe(1);
		expect(tree.state().links[0].target.y).toBe(2);
	});

	test('animates node when moved', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		jest.runTimersToTime(100);

		expect(tree.state().nodes[1].x).toBe(100);
		expect(tree.state().nodes[1].y).toBe(50);

		tree.setProps({
			nodes: [
				nodes[0],
				{
					x: 120,
					y: 80,
					data: {
						name: 'Black'
					}
				}
			]
		});

		jest.runTimersToTime(50);

		expect(tree.state().nodes[1].x).toBe(115);
		expect(tree.state().nodes[1].y).toBe(72.5);

		jest.runTimersToTime(50);

		expect(tree.state().nodes[1].x).toBe(120);
		expect(tree.state().nodes[1].y).toBe(80);
	});

	test('animates link when moved', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		jest.runTimersToTime(100);

		expect(tree.state().links[0].source.x).toBe(1);
		expect(tree.state().links[0].source.y).toBe(2);
		expect(tree.state().links[0].target.x).toBe(100);
		expect(tree.state().links[0].target.y).toBe(50);

		tree.setProps({
			links: [{
				source: {
					x: 5,
					y: 10,
					data: {
						name: 'Colour'
					}
				},
				target: {
					x: 200,
					y: 100,
					data: {
						name: 'Black'
					}
				}
			}]
		});

		jest.runTimersToTime(50);

		expect(tree.state().links[0].source.x).toBe(4);
		expect(tree.state().links[0].source.y).toBe(8);
		expect(tree.state().links[0].target.x).toBe(175);
		expect(tree.state().links[0].target.y).toBe(87.5);

		jest.runTimersToTime(50);

		expect(tree.state().links[0].source.x).toBe(5);
		expect(tree.state().links[0].source.y).toBe(10);
		expect(tree.state().links[0].target.x).toBe(200);
		expect(tree.state().links[0].target.y).toBe(100);
	});

	test('animates node when added', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		tree.setProps({
			nodes: [
				nodes[0],
				nodes[1],
				{
					x: 120,
					y: 80,
					data: {
						name: 'Purple'
					}
				}
			]
		});

		jest.runTimersToTime(50);

		expect(tree.state().nodes[2].x).toBe(90.25);
		expect(tree.state().nodes[2].y).toBe(60.5);

		jest.runTimersToTime(50);

		expect(tree.state().nodes[2].x).toBe(120);
		expect(tree.state().nodes[2].y).toBe(80);
	});

	test('animates node from parent when added', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		jest.runTimersToTime(100);

		tree.setProps({
			nodes: [
				nodes[0],
				{
					x: 100,
					y: 50,
					data: {
						name: 'Black'
					},
					children: [{
						data: {
							name: 'Purple'
						}
					}]
				},
				{
					x: 120,
					y: 80,
					data: {
						name: 'Purple'
					}
				}
			]
		});

		jest.runTimersToTime(50);

		expect(tree.state().nodes[2].x).toBe(115);
		expect(tree.state().nodes[2].y).toBe(72.5);

		jest.runTimersToTime(50);

		expect(tree.state().nodes[2].x).toBe(120);
		expect(tree.state().nodes[2].y).toBe(80);
	});

	test('animates link when added', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		tree.setProps({
			links: [
				links[0],
				{
					source: {
						x: 5,
						y: 10,
						data: {
							name: 'Colour'
						}
					},
					target: {
						x: 200,
						y: 100,
						data: {
							name: 'Purple'
						}
					}
				}
			]
		});

		jest.runTimersToTime(50);

		expect(tree.state().links[1].source.x).toBe(4);
		expect(tree.state().links[1].source.y).toBe(8);
		expect(tree.state().links[1].target.x).toBe(150.25);
		expect(tree.state().links[1].target.y).toBe(75.5);

		jest.runTimersToTime(50);

		expect(tree.state().links[1].source.x).toBe(5);
		expect(tree.state().links[1].source.y).toBe(10);
		expect(tree.state().links[1].target.x).toBe(200);
		expect(tree.state().links[1].target.y).toBe(100);
	});

	test('animates node when removed', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		jest.runTimersToTime(100);

		tree.setProps({
			nodes: [
				nodes[0]
			]
		});

		jest.runTimersToTime(50);

		expect(tree.state().nodes[1].x).toBe(25.75);
		expect(tree.state().nodes[1].y).toBe(14);

		jest.runTimersToTime(50);

		expect(tree.state().nodes.length).toBe(1);
	});

	test('animates link when removed', () => {
		const tree = mount(<Animated {...defaultProps} steps={2} duration={100}/>);

		jest.runTimersToTime(100);

		tree.setProps({
			links: []
		});

		jest.runTimersToTime(50);

		expect(tree.state().links[0].source.x).toBe(75.25);
		expect(tree.state().links[0].source.y).toBe(38);
		expect(tree.state().links[0].target.x).toBe(100);
		expect(tree.state().links[0].target.y).toBe(50);

		jest.runTimersToTime(50);

		expect(tree.state().links.length).toBe(0);
	});

	test('animates from inital value on mount', () => {
		const tree = mount(<Animated {...defaultProps} duration={100}/>);

		expect(tree.state().nodes[1].x).toBe(1);
		expect(tree.state().nodes[1].y).toBe(2);

		jest.runTimersToTime(100);

		expect(tree.state().nodes[1].x).toBe(100);
		expect(tree.state().nodes[1].y).toBe(50);
	});

	test('does nothing when not enabled', () => {
		const tree = mount(<Animated {...defaultProps} animated={false}/>);

		tree.setProps({
			nodes: [
				nodes[0],
				{
					x: 120,
					y: 80,
					data: {
						name: 'Black'
					}
				}
			]
		});


		expect(tree.state().nodes[1].x).toBe(120);
		expect(tree.state().nodes[1].y).toBe(80);

		expect(tree).toMatchSnapshot();
	});

	test('does not animate when props other than nodes or links change', () => {
		const tree = shallow(<Animated {...defaultProps}/>);

		tree.setProps({ nodeOffset: 5 });

		expect(tree).toMatchSnapshot();
	});
});