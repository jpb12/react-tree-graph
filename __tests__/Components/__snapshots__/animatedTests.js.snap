// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Animated> does not animate when props other than nodes or links change 1`] = `
<Container
  animated={true}
  circleProps={Object {}}
  duration={1}
  easing={[Function]}
  gProps={Object {}}
  getChildren={[Function]}
  height={100}
  keyProp="name"
  labelProp="name"
  links={
    Array [
      Object {
        "source": Object {
          "data": Object {
            "name": "Colour",
          },
          "x": 1,
          "y": 2,
        },
        "target": Object {
          "data": Object {
            "name": "Black",
          },
          "x": 1,
          "y": 2,
        },
      },
    ]
  }
  nodeOffset={5}
  nodeRadius={1}
  nodes={
    Array [
      Object {
        "data": Object {
          "name": "Colour",
        },
        "x": 1,
        "y": 2,
      },
      Object {
        "data": Object {
          "name": "Black",
        },
        "x": 1,
        "y": 2,
      },
    ]
  }
  pathProps={Object {}}
  steps={1}
  svgProps={Object {}}
  textProps={Object {}}
  width={100}
/>
`;

exports[`<Animated> does nothing when not enabled 1`] = `
<Animated
  animated={false}
  circleProps={Object {}}
  duration={1}
  easing={[Function]}
  gProps={Object {}}
  getChildren={[Function]}
  height={100}
  keyProp="name"
  labelProp="name"
  links={
    Array [
      Object {
        "source": Object {
          "data": Object {
            "name": "Colour",
          },
          "x": 1,
          "y": 2,
        },
        "target": Object {
          "data": Object {
            "name": "Black",
          },
          "x": 100,
          "y": 50,
        },
      },
    ]
  }
  nodeOffset={1}
  nodeRadius={1}
  nodes={
    Array [
      Object {
        "data": Object {
          "name": "Colour",
        },
        "x": 1,
        "y": 2,
      },
      Object {
        "data": Object {
          "name": "Black",
        },
        "x": 120,
        "y": 80,
      },
    ]
  }
  pathProps={Object {}}
  steps={1}
  svgProps={Object {}}
  textProps={Object {}}
  width={100}
>
  <Container
    animated={false}
    circleProps={Object {}}
    duration={1}
    easing={[Function]}
    gProps={Object {}}
    getChildren={[Function]}
    height={100}
    keyProp="name"
    labelProp="name"
    links={
      Array [
        Object {
          "source": Object {
            "data": Object {
              "name": "Colour",
            },
            "x": 1,
            "y": 2,
          },
          "target": Object {
            "data": Object {
              "name": "Black",
            },
            "x": 100,
            "y": 50,
          },
        },
      ]
    }
    nodeOffset={1}
    nodeRadius={1}
    nodes={
      Array [
        Object {
          "data": Object {
            "name": "Colour",
          },
          "x": 1,
          "y": 2,
        },
        Object {
          "data": Object {
            "name": "Black",
          },
          "x": 120,
          "y": 80,
        },
      ]
    }
    pathProps={Object {}}
    steps={1}
    svgProps={Object {}}
    textProps={Object {}}
    width={100}
  >
    <svg
      height={100}
      width={100}
    >
      <Link
        key="Black"
        keyProp="name"
        pathProps={Object {}}
        source={
          Object {
            "data": Object {
              "name": "Colour",
            },
            "x": 1,
            "y": 2,
          }
        }
        target={
          Object {
            "data": Object {
              "name": "Black",
            },
            "x": 100,
            "y": 50,
          }
        }
        x1={1}
        x2={100}
        y1={2}
        y2={50}
      >
        <path
          d="M2,1C26,1 26,100 50,100"
        />
      </Link>
      <Node
        circleProps={Object {}}
        gProps={Object {}}
        key="Colour"
        keyProp="name"
        labelProp="name"
        name="Colour"
        offset={1}
        radius={1}
        textProps={Object {}}
        x={1}
        y={2}
      >
        <g
          transform="translate(2, 1)"
        >
          <circle
            r={1}
          />
          <text
            dx={1.5}
            dy={1}
          >
            Colour
          </text>
        </g>
      </Node>
      <Node
        circleProps={Object {}}
        gProps={Object {}}
        key="Black"
        keyProp="name"
        labelProp="name"
        name="Black"
        offset={1}
        radius={1}
        textProps={Object {}}
        x={120}
        y={80}
      >
        <g
          transform="translate(80, 120)"
        >
          <circle
            r={1}
          />
          <text
            dx={1.5}
            dy={1}
          >
            Black
          </text>
        </g>
      </Node>
    </svg>
  </Container>
</Animated>
`;

exports[`<Animated> renders correctly and sets initial state 1`] = `
<Container
  animated={true}
  circleProps={Object {}}
  duration={1}
  easing={[Function]}
  gProps={Object {}}
  getChildren={[Function]}
  height={100}
  keyProp="name"
  labelProp="name"
  links={
    Array [
      Object {
        "source": Object {
          "data": Object {
            "name": "Colour",
          },
          "x": 1,
          "y": 2,
        },
        "target": Object {
          "data": Object {
            "name": "Black",
          },
          "x": 1,
          "y": 2,
        },
      },
    ]
  }
  nodeOffset={1}
  nodeRadius={1}
  nodes={
    Array [
      Object {
        "data": Object {
          "name": "Colour",
        },
        "x": 1,
        "y": 2,
      },
      Object {
        "data": Object {
          "name": "Black",
        },
        "x": 1,
        "y": 2,
      },
    ]
  }
  pathProps={Object {}}
  steps={1}
  svgProps={Object {}}
  textProps={Object {}}
  width={100}
/>
`;
